Try and use XDG_RUNTIME_DIR if no host is set.

SingularityCat opened the pull request:
These commits add support for trying `$XDG_RUNTIME_DIR/mpd/socket` first, if no host is specified (either in `mpd_connection_new` or the `MPD_HOST` environment variable. This is the same location MPD will listen on if no bind settings are set.

---

I've refactored `settings.c` and `connection.c`  a bit - `settings.c` now only does explicit setting processing (either via args or envvars), leaving the `mpd_settings` struct fields "unspecified" (NULL or 0), which are later filled in with whatever defaults end up being used in `connection.c`.

There is the potential for issues as there is a small change in behaviour - if no host is specified and no default connections work, `connection->settings->host` will be NULL. A cursory glance at clients using `libmpdclient` suggests most of them assume this can be the case anyway.

MaxKellermann commented:
I like the feature very much, and I was really looking forward to just merge it, but when I saw the actual code, I was disappointed. I found your commits very hard to read, to the point that I still don't understand it.

You keep mixing several things in one commit; just look at the first commit. It pretends to just move a struct, but it also adds documentation. Adding documentation is good, but if the commit message says "Move mpd_settings struct from settings.c to internal.h.", it should do just that. This is a thing I would have accepted, it's trivial, but it was just foreshadowing the more complicated commits that followed.

Worse is the second commit "Move all default handling to 'mpd_connection_new'". This is not only a semantic ABI change which actually breaks applications (not acceptable), but also looks like you tried to make it as complicated as possible (while also adding duplicate code). And while it pretends to just move code around, it adds a new compile-time option called "default_timeout". And it changes the coding style for no reason, look:
```
-       if (settings == NULL)
-               return settings;
+       if (settings == NULL) {
+               return NULL;
+       }
```
I mean wtf man, your diff is already complicated enough because you keep mixing the wrong things in one commit, but why add coding style changes like that on top?

Then code comments like `/* Host is specified, try to connect, then fail. */` - what does that mean? Connect and then fail? Why would I want to fail after connecting to the host? At best, this comment is superfluous because all it does is describe what the code does, instead of why - but it's also worded wrongly and means something different.

That commit also introduces a memory leak bug, because you removed a `mpd_settings_free()` call for no obvious reason. That regression is obstructed by your way of mixing too many changes in one commit.

Oh, and you're trying to add the first `goto` to libmpdclient. No, no, no, no, and again no again!

Now the third commit:
```
+static bool
+mpd_get_xdg_socket_location(char **strp)
```
How does this function work? What is the parameter? It's called "strp", what is a "strp", how do I even pronounce that? You may want it to mean "string pointer", but it's not a string pointer, it's a pointer to a string pointer. What should I pass here, what are valid values? What does the return value mean? Yes, I read the code and I know what you're trying to do, but you chose the most complicated way to implement it, you didn't document it and used an obscure parameter name; all those are unnecessary things that make your code hard to understand.

The fourth commit reformats code comments added by the second commit. But all it really does is fix a regression introduced by the third commit.

Sorry, as much as I'd like to have this feature, but everything is wrong with this PR.

SingularityCat commented:
Hmm, seems I missed the mark here. Sorry about that.

> You keep mixing several things in one commit; just look at the first commit. It pretends to just move a struct, but it also adds documentation.

I guess I though the documentation was trivial and not really worth mentioning, I will be more explicit in commit messages.

> I mean wtf man, your diff is already complicated enough because you keep mixing the wrong things in one commit, but why add coding style changes like that on top?

This is my fault - I made small changes while trying to get this to work, and wrote in the style I'm more familiar with.

> Worse is the second commit "Move all default handling to 'mpd_connection_new'". This is not only a semantic ABI change which actually breaks applications (not acceptable)

Agreed, I neglected the fact that `mpd_settings_new` is part of the public ABI. I have since found a wrapper library that does use this function to learn what the default values are.

> That commit also introduces a memory leak bug, because you removed a mpd_settings_free() call for no obvious reason. 

I don't think that's entirely true - `mpd_settings_new()` is only called once and I changed `mpd_connection_new` to update the struct instead of freeing then reallocating it.

> Oh, and you're trying to add the first goto to libmpdclient.

I expected this to be controversial and completely understand the aversion to it. Will remove.

---

I had originally tried a less invasive set of changes, but the issue I ran into was the "retry" behaviour in `mpd_connection_new`.

If no explicit configuration is present for `settings->host` (and the build supports sockets), `host` gets set to DEFAULT_SOCKET. The logic for falling back to DEFAULT_HOST in this case is to do a string comparison against DEFAULT_SOCKET. If I were to - instead of DEFAULT_SOCKET - set `host` to a string derived from `$XDG_RUNTIME_DIR`, then this fallback logic wouldn't work. This is why I moved the defaults handling into `connection.c`.

A simple approach might be to abandon the idea of trying the XDG location first, and try it only if connecting to `DEFAULT_SOCKET` fails - this can probably be achieved in the smallest set of changes.

However, thinking about this has made me realise: if a user happens to explicitly set MPD_HOST to the exact value of DEFAULT_SOCKET, then the fallback logic would kick in. Is this intentional? It seems like odd behaviour if one explicit value behaves differently to another. Maybe `strcmp`-based approaches should be avoided?

I will try again.

SingularityCat commented:
How about this approach:

`mpd_settings` grows a new field - `host_source`, which is an enum set by `mpd_settings_new` (or one of it's helpers).

The enum looks like this, and tells you where the host setting came from: either it's EXPLICIT, is the DEFAULT_HOST, is the DEFAULT_SOCKET, or is based off the XDG_RUNTIME_DIR.

```
enum mpd_settings_host_source_type {
    MPD_SETTINGS_HOST_SOURCE_EXPLICIT,
    MPD_SETTINGS_HOST_SOURCE_DEFAULT_HOST,
    MPD_SETTINGS_HOST_SOURCE_DEFAULT_SOCKET,
    MPD_SETTINGS_HOST_SOURCE_XDG_RUNTIME_DIR
}
```

(I'm not 100% on `MPD_SETTINGS_HOST_SOURCE_` as a prefix, maybe `MPD_SETTINGS_HOST_SETTING_` or just `MPD_SETTINGS_HOST_`.)

If 'explicit' or 'default_host', no alternatives are tried.
If 'default socket', then 'default host' is tried.
If 'xdg runtime dir', then 'default socket' followed by 'default host' is tried.
